Manual masternode vaults

domob1812 opened the pull request:
This implements a new regtest `mnvaults.py`, which runs through a "masternode vault" workflow with a presigned unvaulting transaction and a presigned masternode broadcast (so that the collateral private key need not be kept around).  For now, all is done "manually" in the test, without specific support for vaulting/unvaulting from `divid`.

This also contains a bunch of tweaks, simplifications and fixes to the general masternode logic and RPC methods that are necessary to make it work.

domob1812 committed:
Tweak sigTime as necessary for delayed broadcast.

When creating a delayed masternode broadcast, check the sigtime against
the collateral confirmation block and make sure that it meets the
network criterion (signature must not be earlier than the timestamp
of the block that confirmed the collateral 15 times).

If the condition is violated, just use "now + one hour" as sigtime.

domob1812 committed:
Regtest for proposed mn vaults.

This adds a new regtest mnvaults.py, which manually implements the
proposal for masternode vaults with a temporary key and a pre-signed
unvault transaction.

It sets up a masternode with a collateral at a temporary address,
for which we need not retain the private key.  Instead we just keep
a pre-signed transaction to withdraw the funds to a specified address
and the signed masternode broadcast in storage, and use them to
first run the masternode and then unvault the funds.

domob1812 commented:
This is mostly ready, except we should make use of custom reward addresses (#44) to make the new regtest more extensive.  Thus that one should probably be merged first.

domob1812 committed:
Allow just refreshing ping in mn broadcast.

broadcaststartmasternode has a feature to refresh the ping's timestamp
if an explicit signature is passed.  This extends that feature to allow
passing "update_ping" in place of the signature, in which case the
signature contained in the mnb will be kept but the ping will still
be refreshed.

Not passing a second argument at all is still a way to just broadcast
the message as-is.

domob1812 commented:
Let's extend the test with custom reward addresses in a follow-up.  The other changes in here are independent of #44, so this is ready to review and merge (should be mostly straight forward).

galpHub commented on divi/src/masternode.cpp:
```diff
@@ -697,7 +697,22 @@ void CMasternodeBroadcastFactory::createWithoutSignatures(
+    /* We have to ensure that the sigtime of the broadcast itself is later
+       than when the collateral gets its 15 confirmations (and it might not
+       even be confirmed once yet).  */
+    const auto* pindexConf = mnbRet.GetCollateralMinConfirmationBlock();
+    if (pindexConf == nullptr) {
+        /* The signature time is accepted on the network up to one hour
+           into the future.  That will also be more than enough to be
+           later than the confirmation block time assuming the collateral
+           is going to be mined "now".  */
+        mnbRet.sigTime = GetAdjustedTime() + 60 * 60;
+        LogPrint("masternode", "Using future sigtime for masternode broadcast: %d\n", mnbRet.sigTime);
+    } else {
+        mnbRet.sigTime = pindexConf->GetBlockTime();
+        LogPrint("masternode", "Using collateral confirmation time for broadcast: %d\n", mnbRet.sigTime);
+    }
```
Reduce the 1h offset to 45 minutes to allow the broadcast to be used right away without potential p2p clock drift issues.

domob1812 replied:
Good point, I've updated the time now to 45 minutes.

galpHub reviewed (commented):
Yep. This one is basically ready. Just need to fix the time offset to avoid any p2p issues with other broadcast use-cases.

domob1812 commented:
Thanks for the review, I've updated the time bump now and also added a fix for a potential deadlock, where we reduce the time held onto `cs_main` to the required minimum.

galpHub commented on divi/src/BlockFactory.cpp:
```diff
@@ -96,7 +95,10 @@ CBlockTemplate* BlockFactory::CreateNewBlock(const CScript& scriptPubKeyIn, bool
 
-    pblocktemplate->previousBlockIndex = chain_.Tip();
+    {
+        LOCK(mainCS_);
```
We may be able to drop the lock altogether. I was doing some digging into it, and since the chain tip is already captured and passed as a parameter downstream, the lock isnt really required. Moreover, since we're checking the chainTip for being non-null, it would function correctly. Do you have any objections to removing this lock altogether?

domob1812 replied:
I think strictly speaking the lock is required, since `CChain::Tip()` performs some operations on `std::vector` (like `operator[]`).  And those are likely not thread-safe, e.g. the vector could just be in the process of being reallocated or something.  My guess is that in practice it will most likely work almost always, but in theory it could lead to weird bugs.

It might be possible to make `CChain::Tip()` thread safe by caching the tip pointer and using e.g. `std::atomic` inside `CChain`, which is something we could do at some point if you are worried about the lock.  But in the current state, I don't think it would be (completely) safe to remove the lock.

In the end, this is your decision, though.  As I said, I don't think it would cause a lot of issues in practice; so if you think we should remove the lock, there are no objections from my side.

domob1812 commented:
Reopened as #81 with fixed branches.
